### PR TITLE
Tdix

### PR DESCRIPTION
**Pull request by tdixono**
Exposing FTParam and gradients in Python.

There is a redundant function declaration in FTP as `ft_eval` since the FT class already has an `eval` function. Do I need to create an FT attribute in the FTP class to solve this?

### CODE REVIEW

File: wrappers/python/c3py/c3py.py [hunk 1]
```diff
@@ -327,6 +327,65 @@ def build_approximation(self, f, fargs, init_rank, verbose, adapt,
 
         self._free_approx_params(c3a, onedopts, low_opts, optnodes)
         c3.fwrap_destroy(fw)
+        
+    def build_reg(self, alg, obj, adaptrank, maxrank, kickrank, 
+                  regweight, roundtol, als_max_sweep, opt_relftol, 
+                  kfold, verbose, kristoffel):
+
```
Comment (goroda) [inline wrappers/python/c3py/c3py.py:334]: Can you add some documentation here to indicate what it returns?

File: pyexamples/c3py_param_example.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+ft_evals = ftp_regress.ft_eval(Xtest)
+err = np.linalg.norm(ft_evals - Ytest) / np.linalg.norm(Ytest)
+mse = np.square(ft_evals - Ytest).mean()
+print("Before Relmse = ", err, " MSE = ", mse)
+
+
+print("--- GRADIENT DESCENT ---")
+params = np.random.randn(nparams)
+print("First Random Updated Parameters: ", params[:2])
```
Comment (goroda) [inline pyexamples/c3py_param_example.py:77]: Should say: "First initialize with random parameters"

File: wrappers/python/c3py/c3py.py [hunk 3]
```diff
@@ -782,8 +811,80 @@ def func(x, param=None):
...
-    
+class FTparam(object):
+
+    def __init__(self, reg):
+        self.reg = reg
+        self.ftp = c3.ft_regress_get_ft_param(reg)
+        self.ft = c3.ft_param_get_ft(self.ftp)
+        self.dim = c3.ft_param_get_dim(self.ftp)
+        self.nparams = self.get_nparams()
+        self.mem = c3.sl_mem_manager_alloc(self.dim,1,self.nparams,c3.LINEAR_ST)
```
Comment (goroda) [inline wrappers/python/c3py/c3py.py:822]: This will only work with c3.LINEAR_ST -- so nonlinear parameterizations will not be allowed. Is this intended?

File: pyexamples/c3py_param_example.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+learning_rate = 5e-3
+
+errors = np.zeros(epochs)
+for e in range(epochs):
+    
+    error_diff = (Y - ftp_regress.ft_eval(X))
+    errors[e] = (1/len(Y))*(error_diff**2).sum()
+    
+    param_grads = np.zeros((len(Y), nparams))
+    for d in range(len(Y)):
```
Comment (goroda) [inline pyexamples/c3py_param_example.py:91]: Why is this a loop? the C code should take a vectorized input/output. I think you should avoid a loop here.

**Comment by tdixon6585** [comment]
@goroda 
I've updated the code based on your comments. Some new updates include adding C functions that evaluate the FT and gradients for multiple data points. Also, the gradient calculations have moved to one FTP function based on if the parameters are linear. 

Side note: 
Based on some C code (sl_mem_manager_check_structure), it appears that all basis function's parameters are linear as long as the kernel center is not being adapted. Is this correct?

**Review by goroda — APPROVED** [review]
